### PR TITLE
[RDY] vim-patch:8.0.{0714, 0720}

### PR DESCRIPTION
**Pull request by gmelodie**
patch 8.0.0714

**Problem:** When a timer causes a command line redraw the " that is displayed for CTRL-R goes missing.
**Solution:** Remember an extra character to display.
https://github.com/vim/vim/commit/a92522fbf3a49d06e08caf010f7d7b0f58d2e131

patch 8.0.0720
    
**Problem:** Unfinished mapping not displayed when running timer.
**Solution:** Also use the extra_char while waiting for a mapping and digraph.
https://github.com/vim/vim/commit/6a77d2667e982655f6adacee774ee7aa2581bd8a

### CODE REVIEW

File: src/nvim/ex_getln.c [hunk 1]
```diff
@@ -1342,6 +1344,7 @@ static int command_line_handle_key(CommandLineState *s)
 
   case Ctrl_R:                        // insert register
     putcmdline('"', true);
+    extra_char = '"';
```
Comment (bfredl) [inline src/nvim/ex_getln.c]: We already have `ccline.special_char` in nvim for this, but it needs to be made unconditional in `putcmdline`

**Comment by janlazo** [comment]
This patch should go together with vim-patch:8.0.0720.

**Comment by justinmk** [comment]
cmdline_speci is failing, maybe it needs to be updated? You can [run the test](https://github.com/neovim/neovim/blob/master/test/README.md#running-tests) like this:

    TEST_FILE=test/functional/ui/cmdline_spec.lua make functionaltest

**Comment by LTKills** [comment]
The failing test is:
```
-- erase information, so we check if it is retransmitted
    command("mode")
    screen:expect{grid=[[
      ^                         |
      {1:~                        }|
      {1:~                        }|
      {1:~                        }|
                               |
    ]], cmdline={{
      firstc = ":",
      content = {{"xx"}},
      pos = 2,
      special = {'"', true}
    }, {
      firstc = "=",
      content = {{"1"}, {"+"}, {"2"}},
      pos = 3,
    }}, reset=true}
```

What is actually being returned is this:

```
cmdline={{
      firstc = ":",
      content = {{"xx"}},
      pos = 2,
      <missing special character here>
    }
```

It seems the `mode` command redraws the screen, I'm not sure if the `<C-R>`should be present after the screen is redrawn.

File: src/nvim/ex_getln.c [hunk 2]
```diff
@@ -1351,7 +1351,8 @@ static int command_line_handle_key(CommandLineState *s)
     if (s->i == Ctrl_R) {
       s->c = plain_vgetc();              // CTRL-R CTRL-R <char>
     }
-    --no_mapping;
+    ccline.special_char = NUL;
```
Comment (bfredl) [inline src/nvim/ex_getln.c]: This is incorrect, as it is scoped in nvim (unlike the original code). This should fix the commented issue also.

**Comment by justinmk** [comment]
LGTM. is this ready?

File: src/nvim/ex_getln.c [hunk 3]
```diff
@@ -1447,6 +1447,9 @@ static int command_line_handle_key(CommandLineState *s)
              && ccline.cmdbuff[ccline.cmdpos - 1] != ' ');
 
     set_cmdspos_cursor();
+    if (ccline.special_char != NUL) {
+      putcmdline(ccline.special_char, ccline.special_shift);
+    }
```
Comment (justinmk) [inline src/nvim/ex_getln.c]: @LTKills  Why do this here, in the `K_LEFT`/`K_S_LEFT`/... case? I don't see anything like that in https://github.com/vim/vim/commit/a92522fbf3a49d06e08caf010f7d7b0f58d2e131 .

**Comment by justinmk** [comment]
merged. https://github.com/neovim/neovim/commit/3b7a4f233bd53eb901904992227be3fffe5e621a

**Pull request closed without merging**